src/__tests__/helpers.js

```diff
@@ -53,42 +49,44 @@ describe('query container validation throws when validation fails', () => {
   })
 })
 
-test('should always use realTimers before using callback when timers are faked with useFakeTimers', () => {
-  const originalSetTimeout = globalObj.setTimeout
+describe('run with real timers', () => {
+  const realSetTimeout = global.setTimeout
 
-  // legacy timers use mocks and do not rely on a clock instance
-  jest.useFakeTimers('legacy')
-  runWithRealTimers(() => {
-    expect(originalSetTimeout).toEqual(globalObj.setTimeout)
+  afterEach(() => {
+    jest.useRealTimers()
+    global.setTimeout = realSetTimeout
```

Comment:
Could you explain with a code comment why we need to restore this? Naively I would expect `jest.useRealTimers` to handle it.

Reply:
I've added comments.

`jest.useRealTimers` restores the timers to the values they had before `jest.useFakeTimers` is called.
If `jest.useFakeTimers` has not been called, it does nothing.

See comments here for a more in depth explanation what happened (before the changes): https://github.com/testing-library/dom-testing-library/blob/24ebf282d10782b3d0a22b939a1b2950d4b5bd1b/src/__tests__/helpers.js#L63-L119

Reply:
> jest.useRealTimers restores the timers to the values they had before jest.useFakeTimers is called.

I would expect jest to restore this i.e. `jest.useFakeTimers().useRealTimers();` should have no effect. It sounds like you're saying that it does which may be a bug with jest.

Reply:
That has no effect. It restores the state before the fake was activated.
That's why we need to restore manually. One of our tests replaces `setTimeout` per assignment and not with `jest.useFakeTimers`.

If you replace a timer function manually between `useFakeTimers` and `useRealTimers`, it breaks `useRealTimers` as shown in that commit.
This might be considered a bug with jest, but could also be considered to be deliberately messing with the constraints jest imposes. ;)
Either way it is solved with the changes here.